sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala

```diff
@@ -61,6 +63,37 @@ abstract class SubqueryExpression(
   }
 }
 
+/**
+ * This expression is used to represent any form of subquery expression namely
+ * ListQuery, Exists and ScalarSubquery. This is only used to make sure the
+ * expression equality works properly when LogicalPlan.sameResult is called
+ * on plans containing SubqueryExpression(s). This is only a transient expression
+ * that only lives in the scope of sameResult function call. In other words, analyzer,
+ * optimizer or planner never sees this expression type during transformation of
+ * plans.
+ */
+case class CanonicalizedSubqueryExpr(expr: SubqueryExpression)
```

Comment:
Do we really need this new abstraction?

Actually I don't think we should compare `SubqueryExpression`s and wonder how to canonicalize them.

Basically, in `sameResult`, by adding new condition `(left.subqueries, right.subqueries).zipped.forall(_ sameResult _)`, we can achieve the same goal and all added tests are passed.